<a href="https://colab.research.google.com/github/garynico/price-prediction/blob/main/UCI_Crimminal_Gary_Nico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# UCI Crimes and Community Dataset
### Gary Nico /  2340081374

## Import Library

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.impute import SimpleImputer
from sklearn.metrics import mean_squared_error

## Import Dataset

In [2]:
df = pd.read_csv('communities.csv')

## 3a. Preprocessing

In [3]:
df.head()

state county  ... PolicBudgPerPop ViolentCrimesPerPop
0      8      ?  ...            0.14                0.20
1     53      ?  ...               ?                0.67
2     24      ?  ...               ?                0.43
3     34      5  ...               ?                0.12
4     42     95  ...               ?                0.03

[5 rows x 128 columns]

In [4]:
df.describe()

state         fold  ...  LemasPctOfficDrugUn  ViolentCrimesPerPop
count  1994.000000  1994.000000  ...          1994.000000          1994.000000
mean     28.683551     5.493982  ...             0.094052             0.237979
std      16.397553     2.873694  ...             0.240328             0.232985
min       1.000000     1.000000  ...             0.000000             0.000000
25%      12.000000     3.000000  ...             0.000000             0.070000
50%      34.000000     5.000000  ...             0.000000             0.150000
75%      42.000000     8.000000  ...             0.000000             0.330000
max      56.000000    10.000000  ...             1.000000             1.000000

[8 rows x 102 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1994 entries, 0 to 1993
Columns: 128 entries, state to ViolentCrimesPerPop
dtypes: float64(100), int64(2), object(26)
memory usage: 1.9+ MB


### Remove Unecessary Column

Kita akan drop kolom (features) yang tidak akan berpengaruh pada proses regression atau classification, seperti:

*   State: kode state
*   County: kode daerah
*   Community: kode komunitas
*   CommunityName: nama komunitas
*   Fold: untuk kepenting cross validation

Semua kolom diatas hanyalah untuk kepentingan informatif / visualisasi tetapi tidak berkontribusi untuk kepentingan regression dan classification, jadi kita bisa drop kolom-kolom tersebut

In [6]:
df = df.drop(columns=['state','county','community','communityname','fold'], axis=1)
df.head()

population  householdsize  ...  PolicBudgPerPop  ViolentCrimesPerPop
0        0.19           0.33  ...             0.14                 0.20
1        0.00           0.16  ...                ?                 0.67
2        0.00           0.42  ...                ?                 0.43
3        0.04           0.77  ...                ?                 0.12
4        0.01           0.55  ...                ?                 0.03

[5 rows x 123 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1994 entries, 0 to 1993
Columns: 123 entries, population to ViolentCrimesPerPop
dtypes: float64(100), object(23)
memory usage: 1.9+ MB


### Handle Missing Values
 Dari 122 kolom ada 23 kolom yang memiliki value null. 

In [8]:
df = df.replace('?', np.nan)
contain_null = df.columns[df.isnull().any()]
print(contain_null)
contain_null.shape

Index(['OtherPerCap', 'LemasSwornFT', 'LemasSwFTPerPop', 'LemasSwFTFieldOps',
       'LemasSwFTFieldPerPop', 'LemasTotalReq', 'LemasTotReqPerPop',
       'PolicReqPerOffic', 'PolicPerPop', 'RacialMatchCommPol',
       'PctPolicWhite', 'PctPolicBlack', 'PctPolicHisp', 'PctPolicAsian',
       'PctPolicMinor', 'OfficAssgnDrugUnits', 'NumKindsDrugsSeiz',
       'PolicAveOTWorked', 'PolicCars', 'PolicOperBudg', 'LemasPctPolicOnPatr',
       'LemasGangUnitDeploy', 'PolicBudgPerPop'],
      dtype='object')


(23,)

Dari 23 kolom yang mempunyai nilai null, 22 kolom memiliki nilai null yang sangat banyak sehingga kita akan drop kolom-kolom tersebut, ada 1 kolom yaitu **OtherPerCap** yang memiliki null values yang sedikit sehingga kita akan mengisi null dengan nilai mean (rata-rata) dari kolom tersebut menggunakan SimpleImputer

In [9]:
imputer = SimpleImputer(missing_values = np.nan, strategy = 'mean')
imputer = imputer.fit(df[['OtherPerCap']])
df[['OtherPerCap']] = imputer.transform(df[['OtherPerCap']])

In [10]:
df = df.dropna(axis=1)
print(df.shape)
df.head()

(1994, 101)


population  householdsize  ...  LemasPctOfficDrugUn  ViolentCrimesPerPop
0        0.19           0.33  ...                 0.32                 0.20
1        0.00           0.16  ...                 0.00                 0.67
2        0.00           0.42  ...                 0.00                 0.43
3        0.04           0.77  ...                 0.00                 0.12
4        0.01           0.55  ...                 0.00                 0.03

[5 rows x 101 columns]

### Find Correlation Feature
Mensortir feature yang paling berpengaruh pada proses regression dan classification

In [11]:
corr = df.corr()

In [12]:
new_corr = corr['ViolentCrimesPerPop'].sort_values(ascending=False)
new_corr.head(10)

ViolentCrimesPerPop    1.000000
PctIlleg               0.737957
racepctblack           0.631264
pctWPubAsst            0.574665
FemalePctDiv           0.556032
TotalPctDiv            0.552777
MalePctDivorce         0.525407
PctPopUnderPov         0.521877
PctUnemployed          0.504235
PctHousNoPhone         0.488243
Name: ViolentCrimesPerPop, dtype: float64

## 3b. Regression

### Import Library
Library yang dipakai untuk proses Regression adalah **Keras**

In [13]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Dropout
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

### Splitting
X = features (100 kolom)

y = label (1 kolom)

In [14]:
data = df.values
X = data[:,0:100]
Y = data[:,100]

### Define and Compile Regression Model
Merancang model regression yang berisi 2 layer

*   Dense Layer 16 + ReLu Activation Function
*   Dense Layer 1




In [15]:
def model():
	model = Sequential()
	model.add(Dense(16, input_dim=100, kernel_initializer='normal', activation='relu'))
	model.add(Dense(1, kernel_initializer='normal'))
	# Compile model
	model.compile(loss='mean_squared_error', optimizer='adam')
	return model

Parameter optimal:

*   Epoch size = 10
*   Batch size = 8


In [16]:
estimators = []
estimators.append(('standardize', StandardScaler()))
estimators.append(('mlp', KerasRegressor(build_fn=model, epochs=10, batch_size=8, verbose=0)))
pipeline = Pipeline(estimators)
kfold = KFold(n_splits=10)
results = cross_val_score(pipeline, X, Y, cv=kfold)
print("Mean: %.2f \nMSE: %f" % (results.mean(), results.std()))

Mean: -0.02 
MSE: 0.002990



Mendapatkan **MSE: 0.00299**. Nilai MSE yang didapatkan  sangat rendah sehingga dapat disimpulkan bahwa kinerja training model sangat baik


## 3c. Classification
Untuk proses klasifikasi, kita harus mengkonversi atribut target (“ViolentCrimesPerPop”) menjadi 3 kelas di kolom **ViolentCrimesPerPopCategory**

### Convert Attribute


1.   Jika ViolentCrimesPerPop < 0.43 maka kelasnya 1
2.   Jika 0.43 <= ViolentCrimesPerPop <= 0.65 kelasnya 2
3.   Jika ViolentCrimesPerPop > 0.65 maka kelasnya 3


In [17]:
df.loc[(df['ViolentCrimesPerPop'] < 0.43 ), 'ViolentCrimesPerPopCategory'] = 1
df.loc[(df['ViolentCrimesPerPop'] >= 0.43) & (df['ViolentCrimesPerPop'] <= 0.67), 'ViolentCrimesPerPopCategory'] = 2
df.loc[(df['ViolentCrimesPerPop'] > 0.67 ), 'ViolentCrimesPerPopCategory'] = 3

df.head()

population  householdsize  ...  ViolentCrimesPerPop  ViolentCrimesPerPopCategory
0        0.19           0.33  ...                 0.20                          1.0
1        0.00           0.16  ...                 0.67                          2.0
2        0.00           0.42  ...                 0.43                          2.0
3        0.04           0.77  ...                 0.12                          1.0
4        0.01           0.55  ...                 0.03                          1.0

[5 rows x 102 columns]

### Splitting
X = features (100 kolom)

y = label (ViolentCrimesPerPopCategory -> low, medium. high)

In [18]:
data = df.values
X = data[:,0:100]
y = data[:,101]

### Define Classification Model
Arsitektur model:

*   Dense layer, Relu Activation
*   Dropout layer
*   Dense Layer
*   Dropout layer
*   Dense layer, Softmax Activation




In [19]:
model = Sequential()
model.add(Dense(64, input_dim=100, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(1, activation='softmax'))


### Compile Classification Model

Parameter optimal:

*   Epoch size = 10
*   Batch size = 16
*   Optimizer = RMSProp



In [20]:
model.compile(loss='binary_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])
model.fit(X, y, epochs=10, batch_size=16, verbose=0)
predictions = np.argmax(model.predict(X), axis=1)
score = model.evaluate(X, y, batch_size=16)

125/125 [==============================] - 0s 1ms/step - loss: -10820.6006 - accuracy: 0.8245



Model mendapatkan **akurasi training sebesar 82%** . Dapat disimpulkan bahwa model memiliki akurasi yang tinggi dan dapat menjalankan proses classification dengan sangat baik.